components/raftstore/src/store/peer.rs

```diff
@@ -1595,6 +1600,26 @@ where
                 "disk_usage" => ?msg.get_disk_usage(),
             );
 
+            for index in msg
+                .get_message()
+                .get_entries()
+                .iter()
+                .map(|e| e.get_index())
+            {
+                while let Some((propose_idx, instant)) = self.pending_propose_instants.front() {
+                    if index == *propose_idx {
+                        ctx.raft_metrics
+                            .proposal_send_wait
+                            .observe(now.saturating_duration_since(*instant).as_secs_f64());
```

Comment:
Does this can affect the performance since it is in the hot path?

Reply:
In my benchmark, the overhead is barely noticable on the flamegraph. It's recorded once per proposal, just like other metrics.

Reply:
Here has two layers of loops, can we reduce it to one layer? e.g. define pending_propose_instants as a hashmap<propose_index, instant>, then we can access it directly.

Reply:
Messages are sent in the same order with proposing. So, this nested loop will not cause quadratic complexity. Therefore, using `VecDeque` (of which the inner implementation is a ring buffer) is actually cheaper than a `HashMap`.

Reply:
Note other metrics use waterfall to reduce performance loss. In the previous benchmark, it can cause about several percents drop. I think you can use poll_ctx.current_time to record the time.